karamel-core/src/main/java/se/kth/karamel/backend/machines/SshMachine.java

```diff
@@ -126,41 +127,108 @@ private boolean anyFailure() {
+                // If we do check, it will lead to deadlock
+                //
+                // After the task gotten the green light to proceed with running the task - parallelism claim
+                // has been satisfied, we check again if the task has failed. While waiting for the claim to be
+                // satisfied, the task might have failed on another node. In that case, we continue the loop without
+                // executing the task. Then we come here, where it is a retry but it has aborted due to failure
+                // and we wait until the failure is resolved
+                if (!retry || hasAbortedDueToFailure) {
+                  logger.debug(String.format("%s: retry: %s hasAborted: %s", machineEntity.getId(),
+                      activeTask.getName(), retry, hasAbortedDueToFailure));
+                  synchronized (rs) {
+                    logger.debug(String.format("%s: Recipe %s RecipeSerializationID: %s Has failed: %s",
+                        activeTask.getMachine().getId(),
+                        activeTask.getName(),
+                        rs.hashCode(),
+                        rs.hasFailed()));
+                    if (rs.hasFailed()) {
+                      logger.info(String.format("%s: Recipe %s has failed on another node. Wait until it succeeds",
+                          machineEntity.getId(), activeTask.getName()));
+                      rs.wait();
+                    } else {
+                      logger.debug(String.format("%s: Recipe %s has NOT failed on another node. Executing",
+                          machineEntity.getId(), activeTask.getName()));
+                    }
+                  }
+                } else {
+                  logger.debug(String.format("%s: %s it is a retry", machineEntity.getId(), activeTask.getName()));
+                }
+              }
+
               prepareSerializedTask(activeTask);
+
+              if (recipeSerialization.isPresent()) {
+                RecipeSerialization rs = recipeSerialization.get();
+                // For explanation read above
+                if (!retry || hasAbortedDueToFailure) {
+                  synchronized (rs) {
+                    logger.debug(String.format("%s: %s Checking again after getting hold of the lock if recipe has " +
+                        "failed", machineEntity.getId(), activeTask.getName()));
+                    if (rs.hasFailed()) {
+                      logger.info(String.format("%s: %s Recipe has failed on another node, releasing the lock and " +
+                              "continue", machineEntity.getId(), activeTask.getName()));
+                      rs.release((RunRecipeTask) activeTask);
+                      logger.debug(String.format("%s: %s Released and continue", machineEntity.getId(),
+                          activeTask.getName()));
+                      hasAbortedDueToFailure = true;
```

Comment:
I'm a bit confused about this part, do we end up in a loop where we will always come back here to under the same condition. Basically, more specific question, when we pause on failure

Reply:
Maybe the confusion comes from the fact that `runTask` does **not** raise an Exception if the recipe has failed. It sets the status to FAILED and returns. The code in question is in an (almost) infinite loop polling tasks from a blocking queue

Reply:
aha, ok